File: FastMines_jfx/src/main/java/fmg/jfx/img/Flag.java
package fmg.jfx.img;

import javafx.geometry.Point2D;
import javafx.scene.canvas.GraphicsContext;
import javafx.scene.paint.Color;

import fmg.common.geom.SizeDouble;
import fmg.core.img.FlagModel;
import fmg.core.img.ImageController;
import fmg.core.img.ImageView;

/** Flag image */
public final class Flag {
    private Flag() {}

    /** Flag image. Base image view JFX implementation
     *  @param <TImage> JFX specific image: {@link javafx.scene.canvas.Canvas} or {@link javafx.scene.image.Image}
     */
    public abstract static class JfxView<TImage> extends ImageView<TImage, FlagModel> {

        public JfxView() {
            super(new FlagModel());
        }

        protected void draw(GraphicsContext g) {
            SizeDouble size = getSize();
            g.clearRect(0,0, size.width, size.height);

            double h = size.height / 100.0;
            double w = size.width  / 100.0;
    //        g.setEffect(new BoxBlur());

            // test
    //        g.setLineWidth(1);
    //        g.setStroke(Color.RED);
    //        g.strokeRect(0, 0, size.width, size.height);

            // perimeter figure points
            Point2D[] p = new Point2D[] {
                new Point2D(13.50 * w, 90 * h),
                new Point2D(17.44 * w, 51 * h),
                new Point2D(21.00 * w, 16 * h),
                new Point2D(85.00 * w, 15 * h),
                new Point2D(81.45 * w, 50 * h)};

            g.setLineWidth(Math.max(1, 7*(w+h)/2));
            g.setStroke(Color.BLACK);
            g.strokeLine(p[0].getX(), p[0].getY(), p[1].getX(), p[1].getY());

            g.setStroke(Color.RED);
            g.beginPath();
            g.moveTo(p[2].getX(), p[2].getY());
            g.bezierCurveTo(95.0 * w,  0 * h,
                            19.3 * w, 32 * h,
                            p[3].getX(), p[3].getY());
            g.bezierCurveTo(77.80 * w, 32.89 * h,
                            88.05 * w, 22.73 * h,
                            p[4].getX(), p[4].getY());
            g.bezierCurveTo(15.83 * w, 67 * h,
                            91.45 * w, 35 * h,
                            p[1].getX(), p[1].getY());
            g.lineTo(p[2].getX(), p[2].getY());
            g.closePath();

            g.stroke();
        }

        @Override
        public void close() {
            super.close();
            getModel().close();
        }

    }

    /////////////////////////////////////////////////////////////////////////////////////////////////////
    //    custom implementations
    /////////////////////////////////////////////////////////////////////////////////////////////////////

    /** Flag image view implementation over {@link javafx.scene.canvas.Canvas} */
    public static class CanvasView extends JfxView<javafx.scene.canvas.Canvas> {

        private CanvasJfx canvas = new CanvasJfx(this);

        @Override
        protected javafx.scene.canvas.Canvas createImage() { return canvas.create(); }

        @Override
        protected void drawBody() { draw(canvas.getGraphics()); }

    }

    /** Flag image view implementation over {@link javafx.scene.image.Image} */
    public static class ImageJfxView extends JfxView<javafx.scene.image.Image> {

        private ImageJfx img = new ImageJfx(this);

        @Override
        protected javafx.scene.image.Image createImage() {
            img.createCanvas();
            return null; // img.createImage(); // fake empty image
        }

        @Override
        protected void drawBody() {
            draw(img.getGraphics());
            setImage(img.createImage()); // real image
        }

    }

    /** Flag image controller implementation for {@link Flag.CanvasView} */
    public static class CanvasController extends ImageController<javafx.scene.canvas.Canvas, Flag.CanvasView, FlagModel> {

        public CanvasController() {
            super(new Flag.CanvasView());
        }

        @Override
        public void close() {
            super.close();
            getView().close();
        }

    }

    /** Flag image controller implementation for {@link Flag.ImageJfxView} */
    public static class ImageJfxController extends ImageController<javafx.scene.image.Image, Flag.ImageJfxView, FlagModel> {

        public ImageJfxController() {
            super(new Flag.ImageJfxView());
        }

        @Override
        public void close() {
            super.close();
            getView().close();
        }

    }

}

File: FastMines_uwp/src/fmg/uwp/img/wbmp/Flag.cs
using System.Linq;
using Windows.UI.Xaml.Media.Imaging;
using Fmg.Common;
using Fmg.Common.Geom;
using Fmg.Uwp.Utils;
using Fmg.Core.Img;

namespace Fmg.Uwp.Img.Wbmp {

    /// <summary> Flag image UWP implementation over <see cref="WriteableBitmap"/> </summary>
    public static class Flag {

        /////////////////////////////////////////////////////////////////////////////////////////////////////
        //    custom implementations
        /////////////////////////////////////////////////////////////////////////////////////////////////////

        /// <summary> Flag image view UWP implementation over <see cref="WriteableBitmap"/> </summary>
        public class WBmpView : ImageView<WriteableBitmap, FlagModel> {

            private WriteableBitmap _bmp;

            public WBmpView()
                : base(new FlagModel())
            { }

            protected override WriteableBitmap CreateImage() {
                var s = Model.Size;
                _bmp = BitmapFactory.New((int)s.Width, (int)s.Height);
                return _bmp;
            }

            protected override void DrawBody() {
                var bmp = _bmp;
                double w = Size.Width  / 100.0;
                double h = Size.Height / 100.0;

                // perimeter figure points
                var p = new[] {
                    new PointDouble(13.5  * w, 90 * h),
                    new PointDouble(17.44 * w, 51 * h),
                    new PointDouble(21.00 * w, 16 * h),
                    new PointDouble(85.00 * w, 15 * h),
                    new PointDouble(81.45 * w, 50 * h)
                };

                bmp.DrawLineAa((int) p[0].X, (int) p[0].Y, (int) p[1].X, (int) p[1].Y, Color.Black.ToWinColor());

                const float tension = 1f; // 0..1
                var clrCurve = Color.Red.ToWinColor();
                bmp.DrawCurve(new[] {
                        p[2],
                        new PointDouble(56.2 * w, 13.3 * h), // new PointDouble(95.0 * w,  0 * h),
                        new PointDouble(54.8 * w, 18.7 * h), // new PointDouble(19.3 * w, 32 * h),
                        p[3]
                    }.PointsAsXyxyxySequence(false).ToArray(), tension, clrCurve);
                bmp.DrawCurve(new[] {
                        p[3],
                        new PointDouble(81.8 * w, 25.0 * h), // new PointDouble(77.80 * w, 32.89 * h),
                        new PointDouble(83.3 * w, 33.5 * h), // new PointDouble(88.05 * w, 22.73 * h),
                        p[4]
                    }.PointsAsXyxyxySequence(false).ToArray(), tension, clrCurve);
                bmp.DrawCurve(new[] {
                        p[4],
                        new PointDouble(51.3 * w, 53.9 * h), // new PointDouble(15.83 * w, 67 * h),
                        new PointDouble(52.7 * w, 48.5 * h), // new PointDouble(91.45 * w, 35 * h),
                        p[1]
                    }.PointsAsXyxyxySequence(false).ToArray(), tension, clrCurve);

                bmp.DrawLineAa((int) p[1].X, (int) p[1].Y, (int) p[2].X, (int) p[2].Y, clrCurve);

            }

            protected override void Disposing() {
                base.Disposing();
                Model.Dispose();
            }

        }

        /// <summary> Flag image controller UWP implementation for <see cref="WBmpView"/> </summary>
        public class WBmpController : ImageController<WriteableBitmap, WBmpView, FlagModel> {

            public WBmpController()
                : base(new WBmpView())
            { }

            protected override void Disposing() {
                base.Disposing();
                View.Dispose();
            }

        }

    }

}
